R/class-spec.R

```diff
@@ -81,27 +81,36 @@ class_friendly <- function(x) {
 }
 
 class_construct <- function(.x, ...) {
-  eval(class_construct_expr(.x, ...))
+  class_constructor(.x)(...)
 }
 
 
-class_construct_expr <- function(.x, ...) {
+class_construct_expr <- function(.x, envir = NULL, package = NULL) {
   f <- class_constructor(.x)
+
+  # If the constructor is an S7 class that is exported from a package, avoid
+  # inlining the full class def instead, inline an expression like
+  # `pkgname::classname()` or `classname()`
+  if (is_class(f) && !is.null(f@package)) {
+    # Check if the class can be resolved as a bare symbol without pkgname::
+    if (identical(package, f@package)) {
```

Comment:
We probably want to update the docs in `?new_class`, `package` arg. Currently, it says:

> Setting package implies that the class is available for external use, so should be accompanied by exporting the constructor.

Reply:
> e.g. when you're extending a non-exported class from the current package?

Whether it's exported or not, using `pkgname::` within package `pkgname` won’t work, because at build time, `::` will throw an error. Therefore, we need to use a bare symbol here.

Reply:
Could you add a comment to that effect? I'm not sure this will be obvious to us (or at least me) later on.